### PR TITLE
Linq Part 3 - Redundant Nodes [API-1624]

### PR DESCRIPTION
**Pull request by emreyigit**
The PR contents steps to remove redundant nodes that are in structured SQL semantic tree. Also, An initial touch for the formatter with its test but rest will be aim of the next part.

### CODE REVIEW

File: src/Hazelcast.Net.Linq.Async/Visitors/RedundantCollector.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            return (select.Where is null && HasSimpleProjection(select))
+                || IsWrapperOfFrom(select);
+        }
+
+        public static bool HasSimpleProjection(SelectExpression select)
+        {
+            foreach (var item in select.Columns)
+            {
+                var column = item.Expression as ColumnExpression;
+                //If column name is changed, so projection to.
```
Comment (zpqrtbnk) [inline src/Hazelcast.Net.Linq.Async/Visitors/RedundantCollector.cs:66]: `too` ;)

**Review by zpqrtbnk — COMMENTED** [review]
I have made a few super-minor comments. I'll approve as soon as you deal with them.

Now... I mostly approve the code because it's clean, documented, with tests etc. I *think* that for the last PR I will spend way more time running all the tests and trying to break everything, but I feel it's too soon.

File: src/Hazelcast.Net.Linq.Async/Visitors/RedundantCollector.cs (hunk 1 above)
Comment (emreyigit) [reply src/Hazelcast.Net.Linq.Async/Visitors/RedundantCollector.cs]: Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/733/commits/a4d0787ab87d2fc40d634997af8dfa2f6264f1e1

File: src/Hazelcast.Net.Linq.Async/Visitors/SubqueryRemover.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text;
+using Hazelcast.Linq.Expressions;
+
+namespace Hazelcast.Linq.Visitors
+{
+    /// <summary>
+    /// Removes(strpies) the provided portions of the given expresession
```
[the remark this answers is not in the document]
Comment (emreyigit) [reply src/Hazelcast.Net.Linq.Async/Visitors/SubqueryRemover.cs]: It should be stripe 😢 Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/733/commits/a4d0787ab87d2fc40d634997af8dfa2f6264f1e1

**Comment by emreyigit** [comment]
On next part,  there will be more tests/cases and bigger coverage 👌

**Review by zpqrtbnk — APPROVED** [review]
Approved!

**Pull request merged**